contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+        sourceContractAddress = sourceContractAddress_;
+
+        tokenReserve = tokenReserve_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "Unauthorized teleporter contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(senderAddress == sourceContractAddress, "Unauthorized Sender");
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        uint256 adjustedAmount = amount;
+        if (tokenReserve > 0) {
+            if (amount > tokenReserve) {
+                emit CollateralAdded({amount: tokenReserve, remaining: 0});
+                adjustedAmount = amount - tokenReserve;
+                tokenReserve = 0;
+            } else {
+                tokenReserve -= amount;
+                emit CollateralAdded({amount: amount, remaining: tokenReserve});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        emit MintNativeTokens(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenMinter-bridgeTokens}.
+     */
+    function transferToSource(
+        address recipient,
+        address feeContractAddress,
+        uint256 feeAmount
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        require(
+            tokenReserve == 0,
+            "Cannot release tokens until source contract is collateralized"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedAmount = 0;
+        if (feeAmount > 0) {
+            adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeContractAddress),
+                feeAmount
+            );
+            IERC20(feeContractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                adjustedAmount
+            );
+        }
```

Comment:
It looks like `feeAmount` and `feeContractAddress` are the relayer fees used by `TeleporterMessenger`. We shouldn't be locking ERC20s in the `NativeTokenSource/Destination` contracts. Instead, we should be doing so in `TeleporterMessenger`. Simply passing these params to `TeleporterMessenger` via the `feeInfo` parameter should be all we need to do.

Reply:
Same scenario in `NativeTokenSource.sol`

Reply:
I'll add this as part of the test so we know we've got the right behaviour

Reply:
Clarified offline. These fees are passed to `TeleporterMessenger` to be used as relayer incentive, but ownership of the ERC20 first needs to be passed to `NativeTokenDestination`.